flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NonBufferResponseDecoderDelegate.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+
+import java.net.ProtocolException;
+
+/**
+ * The parser for messages without specific parser. It receives the whole
+ * messages and then delegate the parsing to the targeted messages.
+ */
+public class NonBufferResponseDecoderDelegate implements NettyMessageDecoderDelegate {
+
+	/** The initial size of the message header cumulator buffer. */
+	private static final int INITIAL_MESSAGE_HEADER_BUFFER_LENGTH = 128;
+
+	/** The cumulation buffer of message header. */
+	private ByteBuf messageBuffer;
+
+	/** The type of messages under processing. */
+	private int msgId = -1;
+
+	/** The length of messages under processing. */
+	private int messageLength;
+
+	@Override
+	public void onChannelActive(ChannelHandlerContext ctx) {
+		messageBuffer = ctx.alloc().directBuffer(INITIAL_MESSAGE_HEADER_BUFFER_LENGTH);
+	}
+
+	@Override
+	public void startParsingMessage(int msgId, int messageLength) {
```

Comment:
Referring to [comment](https://github.com/apache/flink/pull/7368/files#r381871433), we can define the method of `ensureCapacity` which would be called inside `onChannelRead` and the clear method can also be covered inside `onChannelRead`.

Reply:
Have moved the logic into separate `ensureCapacity` method.